design/validation-spi.md

```diff
@@ -0,0 +1,293 @@
+component-specific validations that are directly wired to the surrounding business logic.
+A unified and consistent validation mechanism is currently missing.
+
+There have been several approaches to establish a validation infrastructure, but they have 
+not been followed consistently.
+
+The goal of this proposal is to provide an extensible and unified validation SPI 
+that can be used across the Keycloak code-base. The need for a new validation API was highlighted 
+by the design discussions for the [user-profile](https://github.com/keycloak/keycloak-community/blob/master/design/user-profile.md) support.
+
+The goals of this proposal can be summarized as:
+
+* Provide a uniform and extensible validation SPI
+* Consolidate the scattered validation logic with the new Validation SPI
+
+## Design Principles
+
+The design of the new validation SPI should follow the following design principles:
+
+- Simplicity
+The API should have a small API surface and be easy to use.
+
+- API Granularity
+The API should support low-level validations.
+
+- Efficiency
+Validations should be fast and have little overhead.
+
+- Extensibility
+There should be support for built-in as well as user provided validations.
+The new SPI should be easy to extend, most custom validations should only need one implementation class. 
+Developers should be able to provide reusable custom validation logic components.
+
+- Versatility
+Validations should be flexible and should support validatation of simple values and complex objects.
+It should be possible to parameterize validation executions. Additionally there should be a way to 
+check if the parameterization for a validation is valid.
+Validations should be composable and it should be possible to access other validations from with a validation. 
+
+## Considered Options
+
+### Bean Validation API
+
+The JSR 380 Bean Validation 2.0 is a mature and rich validation framework that is extensively used in Java applications.
+
+- Optimized for validation of complex Java bean objects based on declarative configuration rather than single values.
+
+Allthough, value validations are possible via `javax.validation.Validator#validateValue`, they require a bean class 
+with a property with the annotated constraints.
+```java
+	<T> Set<ConstraintViolation<T>> Validator#validateValue(Class<T> beanType,
+												  String propertyName,
+												  Object value,
+												  Class<?>... groups);
+```
+[Example for validating a value via Validator#validateValue](https://gist.github.com/thomasdarimont/e6e375819b3b1adff86b7669e5b033b4).
+
+- hibernate-validator is currently not used in Keycloak, thus we'd need to add a dependency for:
+```xml
+<dependency>
+    <groupId>org.hibernate</groupId>
+    <artifactId>hibernate-validator</artifactId>
+    <version>6.0.12.Final</version>
+</dependency>
+```
+- No direct support for validation (Constraint) lookups by name. The user-profile support needs a way to lookup validations by name.
```

Comment:
The support for lookup by name is as easy as providing a simple map from name to validator class - just the same as (`ValidatorLookup`) proposed in this PR.

Reply:
I agree with you that you can build this mapping yourself.  I was just looking for a native way to do the lookups.